Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m

```diff
@@ -413,83 +315,187 @@ - (void)loadForDisplay
         self.failedSendBadgeView.tintColor = [UIColor ows_destructiveRedColor];
         [self.contentView addSubview:self.failedSendBadgeView];
 
-        self.payloadConstraints = @[
-            [self.payloadView autoPinLeadingToSuperview],
-            [self.failedSendBadgeView autoPinLeadingToTrailingOfView:self.payloadView],
+        [self.viewConstraints addObjectsFromArray:@[
+            [self.bubbleView autoPinLeadingToSuperview],
+            [self.failedSendBadgeView autoPinLeadingToTrailingOfView:self.bubbleView],
+            [self.failedSendBadgeView autoAlignAxis:ALAxisHorizontal toSameAxisOfView:self.bubbleView],
             [self.failedSendBadgeView autoPinTrailingToSuperview],
-            [self.failedSendBadgeView autoAlignAxis:ALAxisHorizontal toSameAxisOfView:self.payloadView],
             [self.failedSendBadgeView autoSetDimension:ALDimensionWidth toSize:self.failedSendBadgeSize],
             [self.failedSendBadgeView autoSetDimension:ALDimensionHeight toSize:self.failedSendBadgeSize],
-        ];
+        ]];
     } else {
-        self.payloadConstraints = [self.payloadView autoPinWidthToSuperview];
+        [self.viewConstraints addObjectsFromArray:@[
+            [self.bubbleView autoPinLeadingToSuperview],
+            [self.bubbleView autoPinTrailingToSuperview],
+        ]];
     }
 
-    JSQMessagesBubbleImage *_Nullable bubbleImageData;
-    if ([self.viewItem.interaction isKindOfClass:[TSMessage class]]) {
+    if ([self.viewItem.interaction isKindOfClass:[TSMessage class]] && self.hasNonImageBodyContent) {
         TSMessage *message = (TSMessage *)self.viewItem.interaction;
-        bubbleImageData = [self.bubbleFactory bubbleWithMessage:message];
+        self.bubbleView.bubbleColor = [self.bubbleFactory bubbleColorWithMessage:message];
+    } else {
+        // Media-only messages should have no background color; they will fill the bubble's bounds
+        // and we don't want artifacts at the edges.
+        self.bubbleView.bubbleColor = nil;
     }
 
-    self.textBubbleImageView.image = bubbleImageData.messageBubbleImage;
-
     [self updateDateHeader];
     [self updateFooter];
 
```

Comment:
Now that we're going to be vertically stacking up 5-6 views within the bubble, we "create if necessary" each possible using other methods, then arrange them (and space them out) appropriately within this method.